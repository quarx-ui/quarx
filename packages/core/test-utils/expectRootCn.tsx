import React, { FC } from 'react';
import { render } from '@testing-library/react';

export const expectRootCn = <T extends FC<{ className?: string, children: any }>>(Component: T, name: string) => {
    const className = 'test-class-name';
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const { container: { firstChild: rootNode } } = render(<Component className={className} />);
    expect(rootNode).toHaveClass(className);
    expect(rootNode).toHaveClass(name);
};
