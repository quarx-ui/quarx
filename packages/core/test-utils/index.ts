export { expectRootCn } from './expectRootCn';
export * from './testBEM';
