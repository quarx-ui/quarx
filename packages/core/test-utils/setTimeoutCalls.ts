interface FakeSetTimeout {
    mock: {
        calls: Array<[() => void]>; // TODO Доработать типизацию
    };
}

/** При рендере jest использует setTimeout(_flushCallback, 0),
 * из-за чего проверка количества вызовов setTimeout происходит нестабильно.
 *
 * @return массив вызовов setTimeout без _flushCallback */
export const setTimeoutCalls = () => (setTimeout as unknown as FakeSetTimeout)
    .mock.calls
    .filter(([fn]) => !String(fn).includes('_flushCallback'));
