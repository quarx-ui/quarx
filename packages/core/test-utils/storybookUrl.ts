type PropsType = Record<string, string | boolean | number>;

interface UrlOptions {
    component: string;
    story?: string;
    props?: PropsType;
}

const reducePropsToURLString = (props?: PropsType) => {
    if (!props) {
        return '';
    }

    const keysOfProps = Object.keys(props);
    const valuesOfProps = Object.values(props);

    if (keysOfProps.length > 1) {
        return keysOfProps.reduce((prev, cur) => {
            const currentProp = `${cur}:${props[cur]}`;
            return `${prev};${currentProp}`;
        });
    }
    return `${keysOfProps[0]}:${valuesOfProps[0]}`;
};

export function storyUrl(options: UrlOptions) {
    const { component, story = 'sandbox', props } = options;
    const reducedProps = reducePropsToURLString(props);

    return `?path=/story/core-${component}--${story}&args=${reducedProps}`;
}
