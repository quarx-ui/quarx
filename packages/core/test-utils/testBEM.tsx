import React, { ComponentType } from 'react';
import { screen, render, cleanup } from '@testing-library/react';

export const expectPropsInClasses = (
    element: HTMLElement,
    qxClassname?: string,
) => (
    prop: string,
    value?: string | boolean,
) => {
    const qxSelector = qxClassname ?? element.classList[0];

    const valueIsBoolean = typeof value === 'boolean' || value === 'true' || value === 'false';
    const booleanClassname = value === true || value === 'true' ? `${qxSelector}_${prop}` : undefined;

    const expectedClassname = booleanClassname
        ?? (!valueIsBoolean
            ? `${qxSelector}_${prop}_${value}`
            : undefined);

    if (!expectedClassname) {
        return null;
    }

    return expect(element).toHaveClass(expectedClassname);
};

export const expectPropsMapInClasses = (element: HTMLElement, qxClassname?: string) => {
    const expectPropInClasses = expectPropsInClasses(element, qxClassname);

    return (expectedProps: Record<string, string | boolean>) => {
        Object.keys(expectedProps).forEach((prop) => {
            expectPropInClasses(prop, expectedProps[prop]);
        });
    };
};

export const testStyleParams = <
    S extends Record<string, any>,
    P extends Record<string, any> = Record<string, any>,
>(
    Component: ComponentType<P>,
    expectedDefaultParams: S,
    defaultProps: P = {} as P,
    qxClassname?: string,
) => (props: { [key in keyof S]: S[key][] }) => {
    describe('style params', () => {
        describe('default values', () => {
            const testId = 'testStyleParams-default';

            render(
                <Component
                    {...defaultProps}
                    data-testid={testId}
                />,
            );

            const element = screen.getByTestId(testId);

            Object.entries(expectedDefaultParams).forEach(([key, value]) => {
                it(`${key}`, () => {
                    expectPropsInClasses(element, qxClassname)(key, value);
                });
            });

            cleanup();
        });

        Object.entries(props).forEach(([propName, values]) => {
            describe(propName, () => {
                Object.values(values).forEach((value) => {
                    it(`${value}`, () => {
                        const testId = `testStyleParams-${propName}-${value}`;

                        const testProps = { [propName]: value };
                        const { asFragment } = render(
                            <Component
                                {...(defaultProps ?? {})}
                                {...testProps as P}
                                data-testid={testId}
                            />,
                        );

                        const element = screen.getByTestId(testId);
                        expectPropsMapInClasses(element, qxClassname)(testProps as S);
                        expect(asFragment()).toMatchSnapshot();
                    });
                });
            });
        });
    });
};
