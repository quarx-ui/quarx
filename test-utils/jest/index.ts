export { testRootCn } from './testRootCn';
export * from './testBEM';
