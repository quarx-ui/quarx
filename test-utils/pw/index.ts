export {
    compareStorySnapshots,
    compareVariantSnapshots,
    comparePropertySnapshots,
    comparePropertiesMapSnapshots,
} from './compareSnapshots';
