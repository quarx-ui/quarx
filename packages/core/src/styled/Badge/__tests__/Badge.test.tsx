import React from 'react';
import { render, screen } from '@testing-library/react';
import {
    Badge,
    BadgeSize,
    BadgeBorderRadius,
    BadgeColor,
    BadgeType,
    BadgeStyleParams,
    BadgeProps,
} from '@core';
import { expectPropsMapInClasses, testStyleParams } from '@core/test-utils';

interface CheckPropsInClasses {
    size?: BadgeSize,
    borderRadius?: BadgeBorderRadius,
    color?: BadgeColor,
    type?: BadgeType,
    label?: string,
}

const checkPropsInClasses = (propsWithLabel?: CheckPropsInClasses) => {
    const {
        size = 'medium',
        borderRadius = 'max',
        color = 'brand',
        type = 'contained',
        label = 'Статус',
    } = propsWithLabel ?? {};
    const badgeElement = screen.getByText(label);
    const props = { size, borderRadius, color, type };

    expectPropsMapInClasses(badgeElement)(props);
};

describe('Badge', () => {
    testStyleParams<BadgeStyleParams, BadgeProps>(
        Badge,
        { children: 'Статус' },
    )({
        color: ['brand', 'secondary', 'info', 'success', 'warning', 'danger', 'text'],
        size: ['small', 'medium', 'large'],
        borderRadius: ['xSmall', 'small', 'medium', 'large', 'xLarge', 'max'],
        type: ['contained', 'outlined', 'ghosted'],
    });

    it('default with counter', () => {
        const { asFragment } = render(<Badge counter={55}>Статус</Badge>);

        expect(screen.queryByText('Статус')).toBeInTheDocument();
        expect(screen.queryByText('55')).toBeInTheDocument();
        checkPropsInClasses();
        expect(asFragment()).toMatchSnapshot();
    });
});
