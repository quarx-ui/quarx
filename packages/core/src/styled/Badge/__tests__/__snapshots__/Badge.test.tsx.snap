// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge borderRadius large 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_medium QxBadge_borderRadius_large QxBadge_color_brand css-mco4gc-QxBadge-root"
    data-testid="testStyleParams-borderRadius-large"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge borderRadius max 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_medium QxBadge_borderRadius_max QxBadge_color_brand css-16t6hfq-QxBadge-root"
    data-testid="testStyleParams-borderRadius-max"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge borderRadius medium 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_medium QxBadge_borderRadius_medium QxBadge_color_brand css-1l3l1ao-QxBadge-root"
    data-testid="testStyleParams-borderRadius-medium"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge borderRadius small 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_medium QxBadge_borderRadius_small QxBadge_color_brand css-ncc16e-QxBadge-root"
    data-testid="testStyleParams-borderRadius-small"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge borderRadius xLarge 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_medium QxBadge_borderRadius_xLarge QxBadge_color_brand css-1kkvwac-QxBadge-root"
    data-testid="testStyleParams-borderRadius-xLarge"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge borderRadius xSmall 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_medium QxBadge_borderRadius_xSmall QxBadge_color_brand css-10wi4zz-QxBadge-root"
    data-testid="testStyleParams-borderRadius-xSmall"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge color brand 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_medium QxBadge_borderRadius_max QxBadge_color_brand css-16t6hfq-QxBadge-root"
    data-testid="testStyleParams-color-brand"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge color danger 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_medium QxBadge_borderRadius_max QxBadge_color_danger css-10byh6o-QxBadge-root"
    data-testid="testStyleParams-color-danger"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge color info 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_medium QxBadge_borderRadius_max QxBadge_color_info css-1qd392o-QxBadge-root"
    data-testid="testStyleParams-color-info"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge color secondary 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_medium QxBadge_borderRadius_max QxBadge_color_secondary css-16khalj-QxBadge-root"
    data-testid="testStyleParams-color-secondary"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge color success 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_medium QxBadge_borderRadius_max QxBadge_color_success css-pxe1pl-QxBadge-root"
    data-testid="testStyleParams-color-success"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge color text 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_medium QxBadge_borderRadius_max QxBadge_color_text css-7o92zc-QxBadge-root"
    data-testid="testStyleParams-color-text"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge color warning 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_medium QxBadge_borderRadius_max QxBadge_color_warning css-1t31jgi-QxBadge-root"
    data-testid="testStyleParams-color-warning"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge default with counter 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_medium QxBadge_borderRadius_max QxBadge_color_brand css-16t6hfq-QxBadge-root"
  >
    Статус
    <span
      class="QxCounter QxCounter_type_white QxCounter_size_medium QxCounter_color_brand QxBadge-rightItem QxBadge-item css-1pvj4ck-QxCounter-root-QxBadge-rightItem-QxBadge-item"
    >
      55
    </span>
  </div>
</DocumentFragment>
`;

exports[`Badge size large 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_large QxBadge_borderRadius_max QxBadge_color_brand css-1k5us5v-QxBadge-root"
    data-testid="testStyleParams-size-large"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge size medium 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_medium QxBadge_borderRadius_max QxBadge_color_brand css-16t6hfq-QxBadge-root"
    data-testid="testStyleParams-size-medium"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge size small 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_small QxBadge_borderRadius_max QxBadge_color_brand css-hki5s8-QxBadge-root"
    data-testid="testStyleParams-size-small"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge type contained 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_contained QxBadge_size_medium QxBadge_borderRadius_max QxBadge_color_brand css-16t6hfq-QxBadge-root"
    data-testid="testStyleParams-type-contained"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge type ghosted 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_ghosted QxBadge_size_medium QxBadge_borderRadius_max QxBadge_color_brand css-1y2pki1-QxBadge-root"
    data-testid="testStyleParams-type-ghosted"
  >
    Статус
  </div>
</DocumentFragment>
`;

exports[`Badge type outlined 1`] = `
<DocumentFragment>
  <div
    class="QxBadge QxBadge_type_outlined QxBadge_size_medium QxBadge_borderRadius_max QxBadge_color_brand css-1i7i6x8-QxBadge-root"
    data-testid="testStyleParams-type-outlined"
  >
    Статус
  </div>
</DocumentFragment>
`;
